---
title: "Measuring tissue specificity from single cell data with SPICEY"
author: "Georgina Fuentes-PÃ¡ez"
package: SPICEY
output: 
  BiocStyle::html_document:
    fig_width: 7
    fig_height: 5
vignette: >
  %\VignetteIndexEntry{Measuring tissue specificity from single cell data with SPICEY}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include = FALSE}

library(dplyr)

knitr::opts_chunk$set(
    collapse = TRUE,
    comment = "#>",
    eval = TRUE,
    warning = FALSE,
    message = FALSE,
    fig.align = "center",
    out.width = "80%"
)
```

```{r logo, echo=FALSE, eval=TRUE, out.width='10%'}
knitr::include_graphics("../man/figures/logo_spicey.png", dpi = 800)
```

# Introduction

SPICEY is an R package designed to infer cell type specificity from single-cell ATAC-seq and /or RNA-seq data. It supports multiple linking strategies such as:

* Nearest-gene association
* Co-accessibility-based linking

This vignette demonstrates how to use SPICEY with preprocessed GRanges input files. These input files must be prepared in advance (not provided by the package).

# Installation
Install the development version from GitHub:

```{r install, eval=FALSE,  echo=TRUE}
# install.packages("devtools")
devtools::install_github("georginafp/SPICEY")

```

# Input Requirements

The package expects two files:

* ATAC peaks: With metadata including annotation, distanceToTSS, and optionally nearest gene name.
* RNA features: DE genes or expressed genes annotated with gene coordinates and metadata (e.g., logFC, p-value).
* Co-accessibility Links (only for "coaccessibility" mode): A file containing peak-to-peak links inferred from co-accessibility analysis tools like Signac (LinkPeaks()). File must contain the following columns: Peaks1, Peaks2, coaccess, CCAN1, CCAN2

Must be a GInteractions or GRangesList object with metadata columns (e.g., score).

The peaks in the link file should match those in the ATAC object.

Note: The input files are not included in the package. You must preprocess them externally (see below).

# Preprocessing (Overview)

The following code is not part of the package, but shows how the required input files were generated.

## Single cell ATAC-seq data 

Generated by:

* Reading the cell type differential accessible peaks.
* Annotating distance to TSS using ChIPseeker.
* Classifying peaks as promoter or distal.
* Finding the nearest gene using regioneR and plyranges.

```{r scatac, eval=FALSE,  echo=TRUE}

merged_df <- readRDS("../data/ATAC_DAR.rds") |> unlist()
retsi_gr <- regioneR::toGRanges(as.data.frame(merged_df))

# Annotate with TSS distance
anno <- ChIPseeker::annotatePeak(retsi_gr,
  TxDb = TxDb.Hsapiens.UCSC.hg38.knownGene::TxDb.Hsapiens.UCSC.hg38.knownGene,
  verbose = FALSE
)

retsi_gr$distanceToTSS <- data.frame(anno)$distanceToTSS
retsi_gr$annotation <- "Distal"
retsi_gr$annotation[abs(retsi_gr$distanceToTSS) <= 2000] <- "Promoter"

# Add nearest gene name
genes <- biomart_genes()$gr
retsi_gr$nearestGeneSymbol <- genes$external_gene_name[nearest(retsi_gr, promoters(genes, 1, 0))]

saveRDS(retsi_gr, "data/FINAL_ATAC.rds")

```

```{r output-atac, echo=FALSE}

head(readRDS("../data/FINAL_ATAC.rds"))

```


## Single cell RNA-seq data

Generated by:

* Mapping gene symbols to Ensembl IDs.
* Annotating with biotype and genomic coordinates using biomaRt or similar.
* Filtering for protein-coding genes on standard chromosomes.

Embeds metadata from the DE analysis (e.g., logFC, adjusted p-value).

```{r scrna, eval=FALSE, echo=TRUE}

gr_list <- readRDS("../data/RNA_DEG.rds")

gr_df_list <- lapply(gr_list, as.data.frame)
gr_list <- lapply(names(gr_df_list), function(name) {
  df <- gr_df_list[[name]]
  df$symbol <- rownames(df)
  df$ensembl_id <- AnnotationDbi::mapIds(org.Hs.eg.db::org.Hs.eg.db,
    keys = df$symbol,
    column = "ENSEMBL",
    keytype = "SYMBOL",
    multiVals = "first"
  )

  df_annot <- dplyr::left_join(
    df,
    biomart_genes()$df %>% dplyr::select(chromosome_name, start_position, end_position, strand, ensembl_gene_id, gene_biotype),
    by = c("ensembl_id" = "ensembl_gene_id")
  )

  if (nrow(df_annot) == 0) return(NULL)

  df_annot %>%
    dplyr::filter(!is.na(chromosome_name),
                  !is.na(start_position),
                  !is.na(end_position),
                  gene_biotype == "protein_coding",
                  chromosome_name %in% c(as.character(1:22), "X", "Y")) %>%
    dplyr::mutate(chromosome_name = paste0("chr", chromosome_name)) %>%
    makeGRangesFromDataFrame(
      seqnames.field = "chromosome_name",
      start.field = "start_position",
      end.field = "end_position",
      strand.field = "strand",
      keep.extra.columns = TRUE
    ) %>%
    sort()
})

gr_list <- Filter(Negate(is.null), gr_list)
names(gr_list) <- names(gr_df_list)

# Merge into a single GRanges
retsi_rna <- purrr::imap(gr_list, ~ {
  mcols(.x)$cell_type <- .y
  .x
}) %>%
  purrr::reduce(c)

saveRDS(retsi_rna, "data/FINAL_RNA.rds")

```

```{r output-rna, echo=FALSE}

head(readRDS("../data/FINAL_RNA.rds"))

```

# Example: Run SPICEY

We demonstrate how to use SPICEY with the prepared input files.

## Nearest gene mode

```{r spicey-nearest, eval=FALSE,  echo=TRUE}

result_nearest <- run_spicey(
  atac_path = system.file("extdata", "FINAL_ATAC.rds", package = "SPICEY"),
  rna_path = system.file("extdata", "FINAL_RNA.rds", package = "SPICEY"),
  linking_method = "nearest"
)

```

This mode links each peak to the nearest gene TSS.

## Co-accessibility mode

```{r spicey-coaccessibility, eval=FALSE,  echo=TRUE}

result_coacc <- run_spicey(
  atac_path = system.file("extdata", "FINAL_ATAC.rds", package = "SPICEY"),
  rna_path = system.file("extdata", "FINAL_RNA.rds", package = "SPICEY"),
  links_path = system.file("extdata", "COACC_LINKS.rds", package = "SPICEY"),
  linking_method = "coaccessibility"
)

```


This mode uses co-accessibility relationships to link peaks and genes, e.g., from tools like Signac::LinkPeaks().

# Output

Both modes return a GRanges object with:

```{r spicey-outputs}

## ---------------------------------------------------------------------------##
# SPICEY using nearest ---------------------------------------------------------
## ---------------------------------------------------------------------------##
readRDS("../data/SPICEY_nearest.rds") %>% 
  as.data.frame() %>%
  head() %>%
  print()


## ---------------------------------------------------------------------------##
# SPICEY using co-accessibility ------------------------------------------------
## ---------------------------------------------------------------------------##
readRDS("../data/SPICEY_coaccessible.rds") %>% 
  as.data.frame() %>%
  head() %>%
  print()

```

# Visualization example 

```{r plot}

```

# References
