<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Measuring tissue specificity from single cell data with SPICEY • SPICEY</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Measuring tissue specificity from single cell data with SPICEY">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">SPICEY</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.99.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/georginafp/SPICEY"><span class="fa fa-github"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Measuring tissue specificity from single cell data with SPICEY</h1>
                        <h4 data-toc-skip class="author">Georgina
Fuentes-Páez</h4>
            
      
      <small class="dont-index">Source: <a href="https://github.com/georginafp/SPICEY/blob/HEAD/vignettes/SPICEY.Rmd" class="external-link"><code>vignettes/SPICEY.Rmd</code></a></small>
      <div class="d-none name"><code>SPICEY.Rmd</code></div>
    </div>

    
    
<p><img src="../reference/figures/logo_spicey.png" width="10%" style="display: block; margin: auto;"></p>
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>SPICEY is an R package designed to infer cell type specificity from
single-cell ATAC-seq and /or RNA-seq data. It supports multiple linking
strategies such as:</p>
<ul>
<li>Nearest-gene association</li>
<li>Co-accessibility-based linking</li>
</ul>
<p>This vignette demonstrates how to use SPICEY with preprocessed
GRanges input files. These input files must be prepared in advance (not
provided by the package).</p>
</div>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p>Install the development version from GitHub:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># install.packages("devtools")</span></span>
<span><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"georginafp/SPICEY"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="input-requirements">Input Requirements<a class="anchor" aria-label="anchor" href="#input-requirements"></a>
</h2>
<p>The package expects two files:</p>
<ul>
<li>ATAC peaks: With metadata including annotation, distanceToTSS, and
optionally nearest gene name.</li>
<li>RNA features: DE genes or expressed genes annotated with gene
coordinates and metadata (e.g., logFC, p-value).</li>
<li>Co-accessibility Links (only for “coaccessibility” mode): A file
containing peak-to-peak links inferred from co-accessibility analysis
tools like Signac (LinkPeaks()). File must contain the following
columns: Peaks1, Peaks2, coaccess, CCAN1, CCAN2</li>
</ul>
<p>Must be a GInteractions or GRangesList object with metadata columns
(e.g., score).</p>
<p>The peaks in the link file should match those in the ATAC object.</p>
<p>Note: The input files are not included in the package. You must
preprocess them externally (see below).</p>
</div>
<div class="section level2">
<h2 id="preprocessing-overview">Preprocessing (Overview)<a class="anchor" aria-label="anchor" href="#preprocessing-overview"></a>
</h2>
<p>The following code is not part of the package, but shows how the
required input files were generated.</p>
<div class="section level3">
<h3 id="single-cell-atac-seq-data">Single cell ATAC-seq data<a class="anchor" aria-label="anchor" href="#single-cell-atac-seq-data"></a>
</h3>
<p>Generated by:</p>
<ul>
<li>Reading the cell type differential accessible peaks.</li>
<li>Annotating distance to TSS using ChIPseeker.</li>
<li>Classifying peaks as promoter or distal.</li>
<li>Finding the nearest gene using regioneR and plyranges.</li>
</ul>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">merged_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="st">"../data/ATAC_DAR.rds"</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/base/unlist.html" class="external-link">unlist</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">retsi_gr</span> <span class="op">&lt;-</span> <span class="fu">regioneR</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/regioneR/man/toGRanges.html" class="external-link">toGRanges</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="va">merged_df</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Annotate with TSS distance</span></span>
<span><span class="va">anno</span> <span class="op">&lt;-</span> <span class="fu">ChIPseeker</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ChIPseeker/man/annotatePeak.html" class="external-link">annotatePeak</a></span><span class="op">(</span><span class="va">retsi_gr</span>,</span>
<span>  TxDb <span class="op">=</span> <span class="fu">TxDb.Hsapiens.UCSC.hg38.knownGene</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/TxDb.Hsapiens.UCSC.hg38.knownGene/man/package.html" class="external-link">TxDb.Hsapiens.UCSC.hg38.knownGene</a></span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">retsi_gr</span><span class="op">$</span><span class="va">distanceToTSS</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span><span class="va">anno</span><span class="op">)</span><span class="op">$</span><span class="va">distanceToTSS</span></span>
<span><span class="va">retsi_gr</span><span class="op">$</span><span class="va">annotation</span> <span class="op">&lt;-</span> <span class="st">"Distal"</span></span>
<span><span class="va">retsi_gr</span><span class="op">$</span><span class="va">annotation</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">abs</a></span><span class="op">(</span><span class="va">retsi_gr</span><span class="op">$</span><span class="va">distanceToTSS</span><span class="op">)</span> <span class="op">&lt;=</span> <span class="fl">2000</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"Promoter"</span></span>
<span></span>
<span><span class="co"># Add nearest gene name</span></span>
<span><span class="va">genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/biomart_genes.html">biomart_genes</a></span><span class="op">(</span><span class="op">)</span><span class="op">$</span><span class="va">gr</span></span>
<span><span class="va">retsi_gr</span><span class="op">$</span><span class="va">nearestGeneSymbol</span> <span class="op">&lt;-</span> <span class="va">genes</span><span class="op">$</span><span class="va">external_gene_name</span><span class="op">[</span><span class="fu">nearest</span><span class="op">(</span><span class="va">retsi_gr</span>, <span class="fu">promoters</span><span class="op">(</span><span class="va">genes</span>, <span class="fl">1</span>, <span class="fl">0</span><span class="op">)</span><span class="op">)</span><span class="op">]</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">saveRDS</a></span><span class="op">(</span><span class="va">retsi_gr</span>, <span class="st">"data/FINAL_ATAC.rds"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; GRanges object with 6 ranges and 7 metadata columns:</span></span>
<span><span class="co">#&gt;       seqnames              ranges strand | cell_type    p_val_adj        p_val</span></span>
<span><span class="co">#&gt;          &lt;Rle&gt;           &lt;IRanges&gt;  &lt;Rle&gt; |  &lt;factor&gt;    &lt;numeric&gt;    &lt;numeric&gt;</span></span>
<span><span class="co">#&gt;   [1]    chr22   27241597-27243500      * |    Acinar 9.52412e-278 2.04193e-283</span></span>
<span><span class="co">#&gt;   [2]     chr7 142748901-142749963      * |    Acinar 8.50896e-262 1.82429e-267</span></span>
<span><span class="co">#&gt;   [3]    chr10 116553223-116553959      * |    Acinar 1.61617e-251 3.46500e-257</span></span>
<span><span class="co">#&gt;   [4]    chr10 116587488-116588562      * |    Acinar 2.17904e-239 4.67177e-245</span></span>
<span><span class="co">#&gt;   [5]    chr14   66378630-66379738      * |    Acinar 1.04928e-237 2.24961e-243</span></span>
<span><span class="co">#&gt;   [6]    chr17   82909239-82910392      * |    Acinar 7.41158e-234 1.58901e-239</span></span>
<span><span class="co">#&gt;       avg_log2FC distanceToTSS  annotation nearestGeneSymbol</span></span>
<span><span class="co">#&gt;        &lt;numeric&gt;     &lt;numeric&gt; &lt;character&gt;       &lt;character&gt;</span></span>
<span><span class="co">#&gt;   [1]    6.24965        -16463      Distal               MN1</span></span>
<span><span class="co">#&gt;   [2]    6.15114             0    Promoter             PRSS1</span></span>
<span><span class="co">#&gt;   [3]    7.92141          7130      Distal             PNLIP</span></span>
<span><span class="co">#&gt;   [4]    6.81279         -1823    Promoter          PNLIPRP1</span></span>
<span><span class="co">#&gt;   [5]    5.77576        -31966      Distal           CCDC196</span></span>
<span><span class="co">#&gt;   [6]    6.19518           962    Promoter            ZNF750</span></span>
<span><span class="co">#&gt;   -------</span></span>
<span><span class="co">#&gt;   seqinfo: 22 sequences from an unspecified genome; no seqlengths</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="single-cell-rna-seq-data">Single cell RNA-seq data<a class="anchor" aria-label="anchor" href="#single-cell-rna-seq-data"></a>
</h3>
<p>Generated by:</p>
<ul>
<li>Mapping gene symbols to Ensembl IDs.</li>
<li>Annotating with biotype and genomic coordinates using biomaRt or
similar.</li>
<li>Filtering for protein-coding genes on standard chromosomes.</li>
</ul>
<p>Embeds metadata from the DE analysis (e.g., logFC, adjusted
p-value).</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">gr_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="st">"../data/RNA_DEG.rds"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">gr_df_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">gr_list</span>, <span class="va">as.data.frame</span><span class="op">)</span></span>
<span><span class="va">gr_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">gr_df_list</span><span class="op">)</span>, <span class="kw">function</span><span class="op">(</span><span class="va">name</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">df</span> <span class="op">&lt;-</span> <span class="va">gr_df_list</span><span class="op">[[</span><span class="va">name</span><span class="op">]</span><span class="op">]</span></span>
<span>  <span class="va">df</span><span class="op">$</span><span class="va">symbol</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">df</span><span class="op">)</span></span>
<span>  <span class="va">df</span><span class="op">$</span><span class="va">ensembl_id</span> <span class="op">&lt;-</span> <span class="fu">AnnotationDbi</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/AnnotationDbi/man/AnnotationDb-class.html" class="external-link">mapIds</a></span><span class="op">(</span><span class="fu">org.Hs.eg.db</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/org.Hs.eg.db/man/org.Hs.egBASE.html" class="external-link">org.Hs.eg.db</a></span>,</span>
<span>    keys <span class="op">=</span> <span class="va">df</span><span class="op">$</span><span class="va">symbol</span>,</span>
<span>    column <span class="op">=</span> <span class="st">"ENSEMBL"</span>,</span>
<span>    keytype <span class="op">=</span> <span class="st">"SYMBOL"</span>,</span>
<span>    multiVals <span class="op">=</span> <span class="st">"first"</span></span>
<span>  <span class="op">)</span></span>
<span></span>
<span>  <span class="va">df_annot</span> <span class="op">&lt;-</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate-joins.html" class="external-link">left_join</a></span><span class="op">(</span></span>
<span>    <span class="va">df</span>,</span>
<span>    <span class="fu"><a href="../reference/biomart_genes.html">biomart_genes</a></span><span class="op">(</span><span class="op">)</span><span class="op">$</span><span class="va">df</span> <span class="op">%&gt;%</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">chromosome_name</span>, <span class="va">start_position</span>, <span class="va">end_position</span>, <span class="va">strand</span>, <span class="va">ensembl_gene_id</span>, <span class="va">gene_biotype</span><span class="op">)</span>,</span>
<span>    by <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"ensembl_id"</span> <span class="op">=</span> <span class="st">"ensembl_gene_id"</span><span class="op">)</span></span>
<span>  <span class="op">)</span></span>
<span></span>
<span>  <span class="kw">if</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">df_annot</span><span class="op">)</span> <span class="op">==</span> <span class="fl">0</span><span class="op">)</span> <span class="kw"><a href="https://rdrr.io/r/base/function.html" class="external-link">return</a></span><span class="op">(</span><span class="cn">NULL</span><span class="op">)</span></span>
<span></span>
<span>  <span class="va">df_annot</span> <span class="op">%&gt;%</span></span>
<span>    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">chromosome_name</span><span class="op">)</span>,</span>
<span>                  <span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">start_position</span><span class="op">)</span>,</span>
<span>                  <span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">end_position</span><span class="op">)</span>,</span>
<span>                  <span class="va">gene_biotype</span> <span class="op">==</span> <span class="st">"protein_coding"</span>,</span>
<span>                  <span class="va">chromosome_name</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">as.character</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">22</span><span class="op">)</span>, <span class="st">"X"</span>, <span class="st">"Y"</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>chromosome_name <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"chr"</span>, <span class="va">chromosome_name</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>    <span class="fu">makeGRangesFromDataFrame</span><span class="op">(</span></span>
<span>      seqnames.field <span class="op">=</span> <span class="st">"chromosome_name"</span>,</span>
<span>      start.field <span class="op">=</span> <span class="st">"start_position"</span>,</span>
<span>      end.field <span class="op">=</span> <span class="st">"end_position"</span>,</span>
<span>      strand.field <span class="op">=</span> <span class="st">"strand"</span>,</span>
<span>      keep.extra.columns <span class="op">=</span> <span class="cn">TRUE</span></span>
<span>    <span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/sort.html" class="external-link">sort</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span>
<span></span>
<span><span class="va">gr_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/funprog.html" class="external-link">Filter</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/funprog.html" class="external-link">Negate</a></span><span class="op">(</span><span class="va">is.null</span><span class="op">)</span>, <span class="va">gr_list</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">gr_list</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">gr_df_list</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Merge into a single GRanges</span></span>
<span><span class="va">retsi_rna</span> <span class="op">&lt;-</span> <span class="fu">purrr</span><span class="fu">::</span><span class="fu"><a href="https://purrr.tidyverse.org/reference/imap.html" class="external-link">imap</a></span><span class="op">(</span><span class="va">gr_list</span>, <span class="op">~</span> <span class="op">{</span></span>
<span>  <span class="fu">mcols</span><span class="op">(</span><span class="va">.x</span><span class="op">)</span><span class="op">$</span><span class="va">cell_type</span> <span class="op">&lt;-</span> <span class="va">.y</span></span>
<span>  <span class="va">.x</span></span>
<span><span class="op">}</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">purrr</span><span class="fu">::</span><span class="fu"><a href="https://purrr.tidyverse.org/reference/reduce.html" class="external-link">reduce</a></span><span class="op">(</span><span class="va">c</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">saveRDS</a></span><span class="op">(</span><span class="va">retsi_rna</span>, <span class="st">"data/FINAL_RNA.rds"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; GRanges object with 6 ranges and 9 metadata columns:</span></span>
<span><span class="co">#&gt;       seqnames          ranges strand |       p_val avg_log2FC     pct.1</span></span>
<span><span class="co">#&gt;          &lt;Rle&gt;       &lt;IRanges&gt;  &lt;Rle&gt; |   &lt;numeric&gt;  &lt;numeric&gt; &lt;numeric&gt;</span></span>
<span><span class="co">#&gt;   [1]     chr1     65419-71585      + | 8.60381e-01  4.6307389     0.000</span></span>
<span><span class="co">#&gt;   [2]     chr1   923923-944575      + | 6.73442e-14 -0.9545765     0.333</span></span>
<span><span class="co">#&gt;   [3]     chr1   960584-965719      + | 8.46084e-03 -0.0880457     0.352</span></span>
<span><span class="co">#&gt;   [4]     chr1   966482-975865      + | 2.12779e-01  0.1574298     0.218</span></span>
<span><span class="co">#&gt;   [5]     chr1 1001138-1014540      + | 7.88589e-02  0.1096142     0.418</span></span>
<span><span class="co">#&gt;   [6]     chr1 1020120-1056119      + | 3.02868e-08 -0.3550321     0.533</span></span>
<span><span class="co">#&gt;           pct.2   p_val_adj   cell_type      symbol      ensembl_id</span></span>
<span><span class="co">#&gt;       &lt;numeric&gt;   &lt;numeric&gt; &lt;character&gt; &lt;character&gt;     &lt;character&gt;</span></span>
<span><span class="co">#&gt;   [1]     0.000 1.00000e+00      Acinar       OR4F5 ENSG00000186092</span></span>
<span><span class="co">#&gt;   [2]     0.631 1.32129e-09      Acinar      SAMD11 ENSG00000187634</span></span>
<span><span class="co">#&gt;   [3]     0.459 1.00000e+00      Acinar      KLHL17 ENSG00000187961</span></span>
<span><span class="co">#&gt;   [4]     0.271 1.00000e+00      Acinar     PLEKHN1 ENSG00000187583</span></span>
<span><span class="co">#&gt;   [5]     0.545 1.00000e+00      Acinar       ISG15 ENSG00000187608</span></span>
<span><span class="co">#&gt;   [6]     0.768 5.94227e-04      Acinar        AGRN ENSG00000188157</span></span>
<span><span class="co">#&gt;         gene_biotype</span></span>
<span><span class="co">#&gt;          &lt;character&gt;</span></span>
<span><span class="co">#&gt;   [1] protein_coding</span></span>
<span><span class="co">#&gt;   [2] protein_coding</span></span>
<span><span class="co">#&gt;   [3] protein_coding</span></span>
<span><span class="co">#&gt;   [4] protein_coding</span></span>
<span><span class="co">#&gt;   [5] protein_coding</span></span>
<span><span class="co">#&gt;   [6] protein_coding</span></span>
<span><span class="co">#&gt;   -------</span></span>
<span><span class="co">#&gt;   seqinfo: 24 sequences from an unspecified genome; no seqlengths</span></span></code></pre>
</div>
</div>
<div class="section level2">
<h2 id="example-run-spicey">Example: Run SPICEY<a class="anchor" aria-label="anchor" href="#example-run-spicey"></a>
</h2>
<p>We demonstrate how to use SPICEY with the prepared input files.</p>
<div class="section level3">
<h3 id="nearest-gene-mode">Nearest gene mode<a class="anchor" aria-label="anchor" href="#nearest-gene-mode"></a>
</h3>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">result_nearest</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_spicey.html">run_spicey</a></span><span class="op">(</span></span>
<span>  atac_path <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"FINAL_ATAC.rds"</span>, package <span class="op">=</span> <span class="st">"SPICEY"</span><span class="op">)</span>,</span>
<span>  rna_path <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"FINAL_RNA.rds"</span>, package <span class="op">=</span> <span class="st">"SPICEY"</span><span class="op">)</span>,</span>
<span>  linking_method <span class="op">=</span> <span class="st">"nearest"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>This mode links each peak to the nearest gene TSS.</p>
</div>
<div class="section level3">
<h3 id="co-accessibility-mode">Co-accessibility mode<a class="anchor" aria-label="anchor" href="#co-accessibility-mode"></a>
</h3>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">result_coacc</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_spicey.html">run_spicey</a></span><span class="op">(</span></span>
<span>  atac_path <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"FINAL_ATAC.rds"</span>, package <span class="op">=</span> <span class="st">"SPICEY"</span><span class="op">)</span>,</span>
<span>  rna_path <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"FINAL_RNA.rds"</span>, package <span class="op">=</span> <span class="st">"SPICEY"</span><span class="op">)</span>,</span>
<span>  links_path <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"COACC_LINKS.rds"</span>, package <span class="op">=</span> <span class="st">"SPICEY"</span><span class="op">)</span>,</span>
<span>  linking_method <span class="op">=</span> <span class="st">"coaccessibility"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>This mode uses co-accessibility relationships to link peaks and
genes, e.g., from tools like Signac::LinkPeaks().</p>
</div>
</div>
<div class="section level2">
<h2 id="output">Output<a class="anchor" aria-label="anchor" href="#output"></a>
</h2>
<p>Both modes return a GRanges object with:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co">## ---------------------------------------------------------------------------##</span></span>
<span><span class="co"># SPICEY using nearest ---------------------------------------------------------</span></span>
<span><span class="co">## ---------------------------------------------------------------------------##</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="st">"../data/SPICEY_nearest.rds"</span><span class="op">)</span> <span class="op">%&gt;%</span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt;   seqnames     start       end width strand cell_type annotation distanceToTSS</span></span>
<span><span class="co">#&gt; 1    chr22  27241597  27243500  1904      *    Acinar     Distal        -16463</span></span>
<span><span class="co">#&gt; 2     chr7 142748901 142749963  1063      *    Acinar   Promoter             0</span></span>
<span><span class="co">#&gt; 3    chr10 116553223 116553959   737      *    Acinar     Distal          7130</span></span>
<span><span class="co">#&gt; 4    chr10 116587488 116588562  1075      *    Acinar   Promoter         -1823</span></span>
<span><span class="co">#&gt; 5    chr14  66378630  66379738  1109      *    Acinar     Distal        -31966</span></span>
<span><span class="co">#&gt; 6    chr17  82909239  82910392  1154      *    Acinar   Promoter           962</span></span>
<span><span class="co">#&gt;   nearestGeneSymbol                    region    RETSI RETSI_entropy      GETSI</span></span>
<span><span class="co">#&gt; 1               MN1   chr22:27241597-27243500 3.362162  0.0082795495 0.00000000</span></span>
<span><span class="co">#&gt; 2             PRSS1  chr7:142748901-142749963 3.305732  0.0157299376 2.79794028</span></span>
<span><span class="co">#&gt; 3             PNLIP chr10:116553223-116553959 3.267559  0.0004263284 2.84538303</span></span>
<span><span class="co">#&gt; 4          PNLIPRP1 chr10:116587488-116588562 3.220621  0.0017568919 2.54593232</span></span>
<span><span class="co">#&gt; 5           CCDC196   chr14:66378630-66379738 3.213932  0.1060987533         NA</span></span>
<span><span class="co">#&gt; 6            ZNF750   chr17:82909239-82910392 3.198468  0.0114339583 0.01348086</span></span>
<span><span class="co">#&gt;   GETSI_entropy</span></span>
<span><span class="co">#&gt; 1    0.37851735</span></span>
<span><span class="co">#&gt; 2    0.04157385</span></span>
<span><span class="co">#&gt; 3    0.02196533</span></span>
<span><span class="co">#&gt; 4    0.02450432</span></span>
<span><span class="co">#&gt; 5            NA</span></span>
<span><span class="co">#&gt; 6    0.43761550</span></span>
<span></span>
<span></span>
<span><span class="co">## ---------------------------------------------------------------------------##</span></span>
<span><span class="co"># SPICEY using co-accessibility ------------------------------------------------</span></span>
<span><span class="co">## ---------------------------------------------------------------------------##</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="st">"../data/SPICEY_coaccessible.rds"</span><span class="op">)</span> <span class="op">%&gt;%</span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt;   seqnames    start      end width strand cell_type annotation distanceToTSS</span></span>
<span><span class="co">#&gt; 1    chr22 27241597 27243500  1904      *    Acinar     Distal        -16463</span></span>
<span><span class="co">#&gt; 2    chr22 27241597 27243500  1904      *    Acinar     Distal        -16463</span></span>
<span><span class="co">#&gt; 3    chr22 27241597 27243500  1904      *    Acinar     Distal        -16463</span></span>
<span><span class="co">#&gt; 4    chr22 27241597 27243500  1904      *    Acinar     Distal        -16463</span></span>
<span><span class="co">#&gt; 5    chr22 27241597 27243500  1904      *    Acinar     Distal        -16463</span></span>
<span><span class="co">#&gt; 6    chr22 27241597 27243500  1904      *    Acinar     Distal        -16463</span></span>
<span><span class="co">#&gt;   nearestGeneSymbol genes_coacc                  region    RETSI RETSI_entropy</span></span>
<span><span class="co">#&gt; 1               MN1        GRK3 chr22:27241597-27243500 3.362162    0.00827955</span></span>
<span><span class="co">#&gt; 2               MN1      MYO18B chr22:27241597-27243500 3.362162    0.00827955</span></span>
<span><span class="co">#&gt; 3               MN1       SEZ6L chr22:27241597-27243500 3.362162    0.00827955</span></span>
<span><span class="co">#&gt; 4               MN1      ASPHD2 chr22:27241597-27243500 3.362162    0.00827955</span></span>
<span><span class="co">#&gt; 5               MN1        HPS4 chr22:27241597-27243500 3.362162    0.00827955</span></span>
<span><span class="co">#&gt; 6               MN1        SRRD chr22:27241597-27243500 3.362162    0.00827955</span></span>
<span><span class="co">#&gt;        GETSI GETSI_entropy</span></span>
<span><span class="co">#&gt; 1 0.70350739     0.8133780</span></span>
<span><span class="co">#&gt; 2 0.50443406     0.6735259</span></span>
<span><span class="co">#&gt; 3 0.52743075     0.8017803</span></span>
<span><span class="co">#&gt; 4 0.05734870     0.8576141</span></span>
<span><span class="co">#&gt; 5 0.21182449     0.8787387</span></span>
<span><span class="co">#&gt; 6 0.01494051     0.8670795</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="visualization-example">Visualization example<a class="anchor" aria-label="anchor" href="#visualization-example"></a>
</h2>
</div>
<div class="section level2">
<h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Georgina Fuentes-Páez, Mireia Ramos-Rodriguez, Lorenzo Pasquali.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
