<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Measuring tissue specificity from single cell data with SPICEY • SPICEY</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Measuring tissue specificity from single cell data with SPICEY">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">


    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">SPICEY</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.99.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/SPICEY.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../articles/index.html">Articles</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Measuring tissue specificity from single cell
data with SPICEY</h1>
                        <h4 data-toc-skip class="author">Georgina
Fuentes-Páez</h4>
            
      

      <div class="hidden name"><code>SPICEY.Rmd</code></div>

    </div>

    
    
<p><img src="../reference/figures/logo_spicey.png" width="10%" style="display: block; margin: auto;"></p>
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>SPICEY is an R package designed to infer cell type specificity from
single-cell ATAC-seq and /or RNA-seq data. It supports multiple linking
strategies such as:</p>
<ul>
<li>Nearest-gene association</li>
<li>Co-accessibility-based linking</li>
</ul>
<p>This vignette demonstrates how to use SPICEY with preprocessed
GRanges input files. These input files must be prepared in advance (not
provided by the package).</p>
</div>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p>Install the development version from GitHub:</p>
</div>
<div class="section level2">
<h2 id="input-requirements">Input Requirements<a class="anchor" aria-label="anchor" href="#input-requirements"></a>
</h2>
<p>The package expects two files:</p>
<ul>
<li>ATAC peaks: With metadata including annotation, distanceToTSS, and
optionally nearest gene name.</li>
<li>RNA features: DE genes or expressed genes annotated with gene
coordinates and metadata (e.g., logFC, p-value).</li>
<li>Co-accessibility Links (only for “coaccessibility” mode): A file
containing peak-to-peak links inferred from co-accessibility analysis
tools like Signac (LinkPeaks()). File must contain the following
columns: Peaks1, Peaks2, coaccess, CCAN1, CCAN2</li>
</ul>
<p>Must be a GInteractions or GRangesList object with metadata columns
(e.g., score).</p>
<p>The peaks in the link file should match those in the ATAC object.</p>
<p>Note: The input files are not included in the package. You must
preprocess them externally (see below).</p>
</div>
<div class="section level2">
<h2 id="preprocessing-overview">Preprocessing (Overview)<a class="anchor" aria-label="anchor" href="#preprocessing-overview"></a>
</h2>
<p>The following code is not part of the package, but shows how the
required input files were generated.</p>
<div class="section level3">
<h3 id="single-cell-atac-seq-data">Single cell ATAC-seq data<a class="anchor" aria-label="anchor" href="#single-cell-atac-seq-data"></a>
</h3>
<p>Generated by:</p>
<ul>
<li>Reading the cell type differential accessible peaks.</li>
<li>Annotating distance to TSS using ChIPseeker.</li>
<li>Classifying peaks as promoter or distal.</li>
<li>Finding the nearest gene using regioneR and plyranges.</li>
</ul>
<pre><code><span><span class="co">#&gt; GRanges object with 6 ranges and 7 metadata columns:</span></span>
<span><span class="co">#&gt;       seqnames              ranges strand | cell_type    p_val_adj        p_val</span></span>
<span><span class="co">#&gt;          &lt;Rle&gt;           &lt;IRanges&gt;  &lt;Rle&gt; |  &lt;factor&gt;    &lt;numeric&gt;    &lt;numeric&gt;</span></span>
<span><span class="co">#&gt;   [1]    chr22   27241597-27243500      * |    Acinar 9.52412e-278 2.04193e-283</span></span>
<span><span class="co">#&gt;   [2]     chr7 142748901-142749963      * |    Acinar 8.50896e-262 1.82429e-267</span></span>
<span><span class="co">#&gt;   [3]    chr10 116553223-116553959      * |    Acinar 1.61617e-251 3.46500e-257</span></span>
<span><span class="co">#&gt;   [4]    chr10 116587488-116588562      * |    Acinar 2.17904e-239 4.67177e-245</span></span>
<span><span class="co">#&gt;   [5]    chr14   66378630-66379738      * |    Acinar 1.04928e-237 2.24961e-243</span></span>
<span><span class="co">#&gt;   [6]    chr17   82909239-82910392      * |    Acinar 7.41158e-234 1.58901e-239</span></span>
<span><span class="co">#&gt;       avg_log2FC distanceToTSS  annotation nearestGeneSymbol</span></span>
<span><span class="co">#&gt;        &lt;numeric&gt;     &lt;numeric&gt; &lt;character&gt;       &lt;character&gt;</span></span>
<span><span class="co">#&gt;   [1]    6.24965        -16463      Distal               MN1</span></span>
<span><span class="co">#&gt;   [2]    6.15114             0    Promoter             PRSS1</span></span>
<span><span class="co">#&gt;   [3]    7.92141          7130      Distal             PNLIP</span></span>
<span><span class="co">#&gt;   [4]    6.81279         -1823    Promoter          PNLIPRP1</span></span>
<span><span class="co">#&gt;   [5]    5.77576        -31966      Distal           CCDC196</span></span>
<span><span class="co">#&gt;   [6]    6.19518           962    Promoter            ZNF750</span></span>
<span><span class="co">#&gt;   -------</span></span>
<span><span class="co">#&gt;   seqinfo: 22 sequences from an unspecified genome; no seqlengths</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="single-cell-rna-seq-data">Single cell RNA-seq data<a class="anchor" aria-label="anchor" href="#single-cell-rna-seq-data"></a>
</h3>
<p>Generated by: * Mapping gene symbols to Ensembl IDs. * Annotating
with biotype and genomic coordinates using biomaRt or similar. *
Filtering for protein-coding genes on standard chromosomes.</p>
<p>Embeds metadata from the DE analysis (e.g., logFC, adjusted
p-value).</p>
<pre><code><span><span class="co">#&gt; GRanges object with 6 ranges and 9 metadata columns:</span></span>
<span><span class="co">#&gt;       seqnames          ranges strand |       p_val avg_log2FC     pct.1</span></span>
<span><span class="co">#&gt;          &lt;Rle&gt;       &lt;IRanges&gt;  &lt;Rle&gt; |   &lt;numeric&gt;  &lt;numeric&gt; &lt;numeric&gt;</span></span>
<span><span class="co">#&gt;   [1]     chr1     65419-71585      + | 8.60381e-01  4.6307389     0.000</span></span>
<span><span class="co">#&gt;   [2]     chr1   923923-944575      + | 6.73442e-14 -0.9545765     0.333</span></span>
<span><span class="co">#&gt;   [3]     chr1   960584-965719      + | 8.46084e-03 -0.0880457     0.352</span></span>
<span><span class="co">#&gt;   [4]     chr1   966482-975865      + | 2.12779e-01  0.1574298     0.218</span></span>
<span><span class="co">#&gt;   [5]     chr1 1001138-1014540      + | 7.88589e-02  0.1096142     0.418</span></span>
<span><span class="co">#&gt;   [6]     chr1 1020120-1056119      + | 3.02868e-08 -0.3550321     0.533</span></span>
<span><span class="co">#&gt;           pct.2   p_val_adj   cell_type      symbol      ensembl_id</span></span>
<span><span class="co">#&gt;       &lt;numeric&gt;   &lt;numeric&gt; &lt;character&gt; &lt;character&gt;     &lt;character&gt;</span></span>
<span><span class="co">#&gt;   [1]     0.000 1.00000e+00      Acinar       OR4F5 ENSG00000186092</span></span>
<span><span class="co">#&gt;   [2]     0.631 1.32129e-09      Acinar      SAMD11 ENSG00000187634</span></span>
<span><span class="co">#&gt;   [3]     0.459 1.00000e+00      Acinar      KLHL17 ENSG00000187961</span></span>
<span><span class="co">#&gt;   [4]     0.271 1.00000e+00      Acinar     PLEKHN1 ENSG00000187583</span></span>
<span><span class="co">#&gt;   [5]     0.545 1.00000e+00      Acinar       ISG15 ENSG00000187608</span></span>
<span><span class="co">#&gt;   [6]     0.768 5.94227e-04      Acinar        AGRN ENSG00000188157</span></span>
<span><span class="co">#&gt;         gene_biotype</span></span>
<span><span class="co">#&gt;          &lt;character&gt;</span></span>
<span><span class="co">#&gt;   [1] protein_coding</span></span>
<span><span class="co">#&gt;   [2] protein_coding</span></span>
<span><span class="co">#&gt;   [3] protein_coding</span></span>
<span><span class="co">#&gt;   [4] protein_coding</span></span>
<span><span class="co">#&gt;   [5] protein_coding</span></span>
<span><span class="co">#&gt;   [6] protein_coding</span></span>
<span><span class="co">#&gt;   -------</span></span>
<span><span class="co">#&gt;   seqinfo: 24 sequences from an unspecified genome; no seqlengths</span></span></code></pre>
</div>
</div>
<div class="section level2">
<h2 id="example-run-spicey">Example: Run SPICEY<a class="anchor" aria-label="anchor" href="#example-run-spicey"></a>
</h2>
<p>We demonstrate how to use SPICEY with the prepared input files.</p>
<div class="section level3">
<h3 id="nearest-gene-mode">Nearest gene mode<a class="anchor" aria-label="anchor" href="#nearest-gene-mode"></a>
</h3>
<p>This mode links each peak to the nearest gene TSS.</p>
</div>
<div class="section level3">
<h3 id="co-accessibility-mode">Co-accessibility mode<a class="anchor" aria-label="anchor" href="#co-accessibility-mode"></a>
</h3>
<p>This mode uses co-accessibility relationships to link peaks and
genes, e.g., from tools like Signac::LinkPeaks().</p>
</div>
</div>
<div class="section level2">
<h2 id="output">Output<a class="anchor" aria-label="anchor" href="#output"></a>
</h2>
<p>Both modes return a GRanges object with:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co">## ---------------------------------------------------------------------------##</span></span>
<span><span class="co"># SPICEY using nearest ---------------------------------------------------------</span></span>
<span><span class="co">## ---------------------------------------------------------------------------##</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="st">"../data/SPICEY_nearest.rds"</span><span class="op">)</span> <span class="op">%&gt;%</span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt;   seqnames     start       end width strand cell_type annotation distanceToTSS</span></span>
<span><span class="co">#&gt; 1    chr22  27241597  27243500  1904      *    Acinar     Distal        -16463</span></span>
<span><span class="co">#&gt; 2     chr7 142748901 142749963  1063      *    Acinar   Promoter             0</span></span>
<span><span class="co">#&gt; 3    chr10 116553223 116553959   737      *    Acinar     Distal          7130</span></span>
<span><span class="co">#&gt; 4    chr10 116587488 116588562  1075      *    Acinar   Promoter         -1823</span></span>
<span><span class="co">#&gt; 5    chr14  66378630  66379738  1109      *    Acinar     Distal        -31966</span></span>
<span><span class="co">#&gt; 6    chr17  82909239  82910392  1154      *    Acinar   Promoter           962</span></span>
<span><span class="co">#&gt;   nearestGeneSymbol                    region    RETSI RETSI_entropy      GETSI</span></span>
<span><span class="co">#&gt; 1               MN1   chr22:27241597-27243500 3.362162  0.0082795495 0.00000000</span></span>
<span><span class="co">#&gt; 2             PRSS1  chr7:142748901-142749963 3.305732  0.0157299376 2.79794028</span></span>
<span><span class="co">#&gt; 3             PNLIP chr10:116553223-116553959 3.267559  0.0004263284 2.84538303</span></span>
<span><span class="co">#&gt; 4          PNLIPRP1 chr10:116587488-116588562 3.220621  0.0017568919 2.54593232</span></span>
<span><span class="co">#&gt; 5           CCDC196   chr14:66378630-66379738 3.213932  0.1060987533         NA</span></span>
<span><span class="co">#&gt; 6            ZNF750   chr17:82909239-82910392 3.198468  0.0114339583 0.01348086</span></span>
<span><span class="co">#&gt;   GETSI_entropy</span></span>
<span><span class="co">#&gt; 1    0.37851735</span></span>
<span><span class="co">#&gt; 2    0.04157385</span></span>
<span><span class="co">#&gt; 3    0.02196533</span></span>
<span><span class="co">#&gt; 4    0.02450432</span></span>
<span><span class="co">#&gt; 5            NA</span></span>
<span><span class="co">#&gt; 6    0.43761550</span></span>
<span></span>
<span></span>
<span><span class="co">## ---------------------------------------------------------------------------##</span></span>
<span><span class="co"># SPICEY using co-accessibility ------------------------------------------------</span></span>
<span><span class="co">## ---------------------------------------------------------------------------##</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="st">"../data/SPICEY_coaccessible.rds"</span><span class="op">)</span> <span class="op">%&gt;%</span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt;   seqnames    start      end width strand cell_type annotation distanceToTSS</span></span>
<span><span class="co">#&gt; 1    chr22 27241597 27243500  1904      *    Acinar     Distal        -16463</span></span>
<span><span class="co">#&gt; 2    chr22 27241597 27243500  1904      *    Acinar     Distal        -16463</span></span>
<span><span class="co">#&gt; 3    chr22 27241597 27243500  1904      *    Acinar     Distal        -16463</span></span>
<span><span class="co">#&gt; 4    chr22 27241597 27243500  1904      *    Acinar     Distal        -16463</span></span>
<span><span class="co">#&gt; 5    chr22 27241597 27243500  1904      *    Acinar     Distal        -16463</span></span>
<span><span class="co">#&gt; 6    chr22 27241597 27243500  1904      *    Acinar     Distal        -16463</span></span>
<span><span class="co">#&gt;   nearestGeneSymbol genes_coacc                  region    RETSI RETSI_entropy</span></span>
<span><span class="co">#&gt; 1               MN1        GRK3 chr22:27241597-27243500 3.362162    0.00827955</span></span>
<span><span class="co">#&gt; 2               MN1      MYO18B chr22:27241597-27243500 3.362162    0.00827955</span></span>
<span><span class="co">#&gt; 3               MN1       SEZ6L chr22:27241597-27243500 3.362162    0.00827955</span></span>
<span><span class="co">#&gt; 4               MN1      ASPHD2 chr22:27241597-27243500 3.362162    0.00827955</span></span>
<span><span class="co">#&gt; 5               MN1        HPS4 chr22:27241597-27243500 3.362162    0.00827955</span></span>
<span><span class="co">#&gt; 6               MN1        SRRD chr22:27241597-27243500 3.362162    0.00827955</span></span>
<span><span class="co">#&gt;        GETSI GETSI_entropy</span></span>
<span><span class="co">#&gt; 1 0.70350739     0.8133780</span></span>
<span><span class="co">#&gt; 2 0.50443406     0.6735259</span></span>
<span><span class="co">#&gt; 3 0.52743075     0.8017803</span></span>
<span><span class="co">#&gt; 4 0.05734870     0.8576141</span></span>
<span><span class="co">#&gt; 5 0.21182449     0.8787387</span></span>
<span><span class="co">#&gt; 6 0.01494051     0.8670795</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="visualization-example">Visualization example<a class="anchor" aria-label="anchor" href="#visualization-example"></a>
</h2>
</div>
<div class="section level2">
<h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

      </div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Georgina Fuentes-Páez, Mireia Ramos-Rodriguez, Lorenzo Pasquali.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

      </footer>
</div>






  </body>
</html>
